docs/design/proposals/VRG-Reconciliation.md

```diff
@@ -0,0 +1,132 @@
+# VRG reconciler workflow
+
+This design document covers the VRG CRD and its reconciliation workflow.
+
+**NOTE**: This is added to aid understanding and ease development of the
+various parts in parallel (where possible) and would possibly be merged once
+it reaches initial completion.
+
+## Terminology
+
+- VRG: VolumeReplicationGroup
+- VR: VolumeReplication
+- PVC: PersistentVolumeClaim
+- PV: PersistentVolume
+
+## VRG CRD
+
+- Spec:
+    - PVCSelector
+        - Label selector to identify all the PVCs that are in this group that
+        needs to be replicated to the peer cluster.
+    - VolumeReplicationClass
+        - ReplicationClass of all volumes in this replication group; this
+        value is propagated to children VolumeReplication CRs
+    - ReplicationState
+        - Desired state of all volumes [`primary` or `secondary`] in this
+        replication group; this value is propagated to children
+        VolumeReplication CRs
+    - S3Endpoint
+        - S3 Endpoint to replicate PV metadata
+    - S3SecretName
+        - S3 Endpoint to replicate PV metadata
+- Status:
+    - **TBD**
+
+### VRG Example
+
+```
+---
+apiVersion: ramendr.openshift.io/v1alpha1
+kind: VolumeReplicationGroup
+metadata:
+  name: vrg-sample
+spec:
+  pvcSelector:
+    matchLabels:
+      appname: myapp
+  volumeReplicationClass: "vr-class-sample"
+  replicationState: "Primary"
+  s3Endpoint: http://objectstore.svc:9000
+  s3SecretName: vrg-objectstore-secret
+```
+
+## Reconciling VRG as Primary
+
+1. Select all PVCs that match `PVCSelector` for processing
+1. For each PVC
+    1. If not bound, continue but requeue
+    1. Check `vr-protected` finalizer, if found,
+        1. Check and remove (stale) `vr-protecting` finalizer
+        1. Check and update VR as `primary` and continue
```

Comment:
I am assuming you are referring to `2.i.b` [here](https://github.com/RamenDR/ramen/blob/6be6db75e8b93d9c131b9cf950efb87056961172/docs/design/proposals/VRG-Reconciliation.md#reconciling-vrg-as-primary).

So the intention is that we found the `vr-protected` finalizer, and hence are done creating the VR etc. and can now change VR to desired state.

We will determine actual state of the VR in the next loop `3.i`.

So at this point we have added the said finalizer in a prior reconcile action.

Am I missing something?

Reply:
In the link you posted ([here](https://github.com/RamenDR/ramen/blob/6be6db75e8b93d9c131b9cf950efb87056961172/docs/design/proposals/VRG-Reconciliation.md#reconciling-vrg-as-primary)), there is no 2.i.b.  I believe I was suggest reordering the checks, so that the pseudocode first checks for vr-protecting to perform some action and then checks if vr-protected to do some action.  This doesn't change the pseudo code but hopefully makes it easier to read.